numpy/random/src/distributions/distributions.c

```diff
@@ -43,33 +43,29 @@ void random_standard_uniform_fill_f(bitgen_t *bitgen_state, npy_intp cnt, float
   }
 }
 
-static double standard_exponential_unlikely(bitgen_t *bitgen_state,
-                                                uint8_t idx, double x) {
-  if (idx == 0) {
-    /* Switch to 1.0 - U to avoid log(0.0), see GH 13361 */
-    return ziggurat_exp_r - npy_log1p(-next_double(bitgen_state));
-  } else if ((fe_double[idx - 1] - fe_double[idx]) * next_double(bitgen_state) +
-                 fe_double[idx] <
-             exp(-x)) {
-    return x;
-  } else {
-    return random_standard_exponential(bitgen_state);
-  }
-}
-
 double random_standard_exponential(bitgen_t *bitgen_state) {
   uint64_t ri;
   uint8_t idx;
   double x;
-  ri = next_uint64(bitgen_state);
-  ri >>= 3;
-  idx = ri & 0xFF;
-  ri >>= 8;
-  x = ri * we_double[idx];
-  if (ri < ke_double[idx]) {
-    return x; /* 98.9% of the time we return here 1st try */
+  double shift = 0;
+  while(1) {
```

Comment:
Is the performance of a `while(1)` better than tail recursion?

Reply:
No it's just that the three cases in the if statements are just one-liners, so I thought it was more readable to have everything in a single function (also because I was mirroring boost.. now that you mentioned that this was inspired by the Julia implementation I can see where it came from). Especially with the suggested change, it's perhaps easier to understand what the `shift` represents if everything is in one place.

Anyways I think it was not a good idea from my side to submit two patches, because the change that I was interested in discussing with you developers is better seen in the first patch. This second patch with the refactoring might have created some confusion, I can delete that in case